src/utils/tcr.js

```diff
@@ -0,0 +1,99 @@
+const executePending = async ({
+  batchedSend,
+  fromBlock,
+  tcrContract,
+  toBN,
+  type
+}) => {
+  const statusChangeEvents = await tcrContract.getPastEvents(
+    `${type}StatusChange`,
+    { fromBlock }
+  )
+
+  // Take most recent event for each item.
+  const latestStatusChanges = {}
+  statusChangeEvents.forEach(event => {
+    const itemKey = event.raw.topics[3]
+    if (!latestStatusChanges[itemKey]) {
+      latestStatusChanges[itemKey] = event
+      return
+    }
+    if (event.blockNumber > latestStatusChanges[itemKey].blockNumber)
+      latestStatusChanges[itemKey] = event
+  })
+
+  const undisputedRequests = Object.keys(latestStatusChanges)
+    .map(itemKey => latestStatusChanges[itemKey])
+    .filter(requestEvent => {
+      // Only take status change events emitted when starting a request.
+      const {
+        returnValues: { _status }
+      } = requestEvent
+      return Number(_status) > 1 // Statuses 2 and 3 are registration and removal requested.
+    })
+    .filter(requestEvent => {
+      // Only take requests that were never challenged.
+      const {
+        raw: { topics }
+      } = requestEvent
+      const challenger = topics[2]
+      return (
+        challenger ===
+        '0x0000000000000000000000000000000000000000000000000000000000000000'
+      )
+    })
+    .map(requestEvent =>
+      type === 'Token'
+        ? requestEvent.raw.topics[3]
+        : `0x${requestEvent.raw.topics[3].slice(
+            // Cast from bytes32
+            26,
+            requestEvent.raw.topics[3].length
+          )}`
+    )
+
+  const challengePeriodDuration = toBN(
+    await tcrContract.methods.challengePeriodDuration().call()
+  )
+
+  // Only take submissions that passed the challenge period.
+  const timedOutRequests = (await Promise.all(
+    undisputedRequests.map(async itemKey => {
```

Comment:
You can use `filter` from the get go. I see no reason to do `map` to `filter` to `map`. Just originally filter out requests that are timedOut

Reply:
That was my initial attempt but we run into problems as [filter is synchronous and does not support promises](https://stackoverflow.com/a/33362966/1772948). So I use `map` to fetch the required data first.